presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -107,7 +120,59 @@ public void process(Iterable<QueryExecution> queries)
         return ImmutableMap.copyOf(pools);
     }
 
-    private void updateNodes()
+    private MemoryPoolAssignmentsRequest updateAssignments(Iterable<QueryExecution> queries)
+    {
+        ClusterMemoryPool reservedPool = pools.get(RESERVED_POOL);
+        long version = memoryPoolAssignmentsVersion.incrementAndGet();
+        if (reservedPool != null && allAssignmentsHavePropagated(queries)) {
```

Comment:
I'm not sure why we have the `allAssignmentsHavePropagated` check here?  Is there a problem is some nodes don't see an update?  If so, I don't think your code handles the case where nodes flip their offline/online state.